llvm/lib/CAS/BuiltinCAS.h

```diff
@@ -107,15 +107,38 @@ inline ArrayRef<char> toArrayRef(StringRef Data) {
 }
 
 /// Current hash type for the internal CAS.
-using HasherT = SHA1;
+///
+/// FIXME: This should be configurable via an enum, with at the following values:
```

Comment:
What's the rationale for making it configurable? What situations would demand 16-byte or 32-byte hashes, respectively?

Reply:
Making it configurable is important for:
- Future proofing, to allow us to eventually move to BLAKE7.
- Optimizing memory usage and storage. My reading of the BLAKE3 paper (which could be faulty) gave me the impression that 16B would be a good default for the in-memory CAS and likely many local use cases... seemed to imply it was already as good as or better than SHA-1 (which is what Git uses). Maybe we want an in-memory-only CAS to default to 16B; maybe we want everything to default to 16B; maybe a vendor choice, or a use-case choice made by the build system (using this on-disk CAS for a single build folder? 16B. Using this one for all clang/LLVM on the machine? 24B. Hooking it up to a massive distributed system? 32B).

Reply:
Two follow-ups, depending on your response:
- If you agree, I'll add the rationale to the commit.
- Maybe we should use 16B or 20B now — WDYT?

Reply:
> Making it configurable is important for:
> * Future proofing, to allow us to eventually move to BLAKE7.

Ah sorry, I should have been more specific that I was responding more to the idea of configuring between blake3 and blake3_16, since those are the cases you listed.  I agree that we would want to be able to handle a new better hash in the future, but it isn't obvious to me that we would actually want to add a _parameter_ for that until we have such a hash function in hand, since the API is unstable anyway.

> Optimizing memory usage and storage.

Maybe this is obvious to others, but where do we materialize CASIDs as full hashes rather than using indirection? How important would the saving be?

> My reading of the BLAKE3 paper... Maybe we want an in-memory-only CAS to default to 16B; maybe we want everything to default to 16B; maybe a vendor choice, or a use-case choice made by the build system ...

If there is a significant benefit to using smaller hashes for the in-memory CAS, and it has good enough hash quality for that use case, that seems like a fine reason to make it configurable to me.

> maybe we want everything to default to 16B
> Maybe we should use 16B or 20B now — WDYT?

If we feel that 32B is necessary for some cases, I would keep 32B as the default for the on-disk format for now.

Reply:
> > Optimizing memory usage and storage.
> 
> Maybe this is obvious to others, but where do we materialize CASIDs as full hashes rather than using indirection? How important would the saving be?

- They are stored in the CAS itself, one per object. Saving 16B per object if you have 1M objects saves 16MB. I don't know if this matters, but this is what I was referring to previously. Also:
- Any non-integrated action cache needs to store the IDs of the action results that are cached.
- An action cache that lives longer than the CAS also needs to recursively store any objects. If we support adapting a generic hash map (where the object graph is opaque), it won't be able to optimize the CAS references.

The latter two would play out something like this:
```
/// A generic map interface.
class GenericMap {
  /// Insert into the map, calling ValueWriter if and only if Key is new.
  virtual Error put(const Twine &Key, function_ref<Error (raw_ostream &)> ValueWriter) = 0;
  virtual Error read(const Twine &Key, raw_ostream &OS) const = 0;
  virtual Error poison(const Twine &Key) = 0;
};

/// Adapt a generic map to cache actions. Assume result objects are stored elsewhere.
class MapToActionCacheAdaptor : public ActionCache {
  Map &M;
public:
  Error put(const ActionKey &Key, const CASID &Result) override {
    return M.put(Key.toString(), [&]() { return Result.toString()}).takeError();
  }
  // ...
};

/// Adapt a generic map to cache actions. Store result objects.
class MapToActionCacheAndStorageAdaptor : public ActionCache {
  Map &M;
  ObjectStore &CAS;

public:
  Error put(const ActionKey &Key, const CASID &Result) {
    SmallVector<CASID> Worklist = {Result};
    DenseMap<CASID> Visited;
    auto addChildToWorklist = [&](const CASID &ID) {
      if (Visited.insert(CASID).second)
        Worklist.push_back(Child);
    };

    while (!Worklist.empty()) {
      CASID ID = Worklist.pop_back_val();
      auto serializeObject = [&](raw_ostream &OS) -> Error {
        return CAS.serializeObject(OS, ID, addChildToWorklist));
      };

      // If the object is already there, serializeObject won't be called, and we
      // don't need to descend.
      if (Error E = M.put(StringRef("object:") + Result.toString(), serializeObject).moveInto(WasInserted))
        return E;
    }
    return M.put(StringRef("action:") + Key.toString(), Result.toString());
  }
  // ...
};
```
The `serializeObject` for a node would be something like:
```
endian::writer EW(OS, endian::little);
EW.write<uint32_t>(Obj.getNumRefs());
EW.write<uint32_t>(CAS.getHashSize());
for (const CASID &Child : Obj.children()) {
  EW << Child.toString();
  addChildToWorklist(Child);
}
EW.write<uint64_t>(CAS.getData().size());
EW.write(Obj.getData());
```
So each CAS reference here ends up costing the full amount in the object that references it.

... That said, maybe this says more about `MapToActionCacheAndStorageAdaptor`'s suitability as a plugin interface than anything else.